Source/BlackHoles/PunctureTracker.cpp

```diff
@@ -0,0 +1,235 @@
+/* GRChombo
+ * Copyright 2012 The GRChombo collaboration.
+ * Please refer to LICENSE in GRChombo's root directory.
+ */
+
+#include "PunctureTracker.hpp"
+#include "ChomboParameters.hpp" // for writing data
+#include "DimensionDefinitions.hpp"
+#include "InterpolationQuery.hpp"
+#include "SmallDataIO.hpp"   // for writing data
+#include "UserVariables.hpp" // for writing data
+
+//! Set punctures post restart
+void PunctureTracker::initial_setup(
+    const std::vector<std::array<double, CH_SPACEDIM>> &initial_puncture_coords,
+    const std::string &a_checkpoint_prefix)
+{
+    m_punctures_filename = a_checkpoint_prefix + "Punctures";
+
+    // first set the puncture data
+    // m_num_punctures is only set later
+    m_puncture_coords = initial_puncture_coords;
+}
+
+void PunctureTracker::set_interpolator(
+    AMRInterpolator<Lagrange<4>> *a_interpolator)
+{
+    change_interpolator(a_interpolator);
+
+    int current_step = m_interpolator->getAMR().s_step;
+
+    if (current_step == 0)
+    {
+        // if it is the first timestep, use the param values
+        // rather than look for the output file, e.g. for when
+        // restart from IC solver checkpoint
+        set_initial_punctures();
+    }
+    else
+    {
+        // look for the current puncture location in the
+        // puncture output file (it needs to exist!)
+        read_in_punctures(current_step,
+                          m_interpolator->getAMR().getCurrentTime());
+    }
+}
```

Comment:
I don't like the fact that the name of this function doesn't describe what most of it is doing. Can we rename the existing `change_interpolator` function to `set_interpolator` (also consistent with `GRAMR::set_interpolator`) and rename this function to `restart_punctures` (or similar) and remove the code from it that sets/changes the interpolator.